---
title: "Introduction to differential expression analysis for bulk-seq data"
author: "Amanda Zacharias"
date: "`r format(Sys.time(), '%B %d %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr) # 1.45
library(kableExtra) # 1.3.4
```

This page provides an introduction to performing differential expression analysis (DEA) with DESeq2
The data used for this analysis comes from <a href="https://doi.org/10.1038/s41598-017-16664-z">Lopes <em>et al.</em>'s study</a> on sex differences of pain response in mice.

# Environment

This tutorial is written in R version 4.2.1. 

Packages: 
```{r message=FALSE, warning=FALSE, error=FALSE}
library(readxl) # version 1.4.3
library(GEOquery) # version 2.66.0
library(dplyr) # version 1.1.4
library(DESeq2) # version 1.38.3
library(ggplot2) # version 3.4.4
library(cowplot) # version 1.1.2
```

# Folder system setup

Make folder for project files

- `file.path()` combines paths/strings together in a way that is both Windows and Mac compatible
- `getwd()` retrieves the current working directory, from which all files are stored
- `system()` runs code in Unix while still being able to use R functions
```{r}
projDir <- file.path(getwd(), "deaDemo")
# Inputs 
dataDir <- file.path(projDir, "0_data")

# Outputs
```

```{r eval=FALSE}
# Make folders
system(paste('mkdir', projDir, dataDir))
```

# Tutorial data

## Download and load raw gene counts
```{r eval=FALSE}
# Download
system(paste('wget -P', dataDir, "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100035/suppl/GSE100035%5FSupplementary%5FTable%5F1.xlsx"))
system(paste('wget -P', dataDir, "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100035/suppl/GSE100035%5FSupplementary%5FTable%5F2.xlsx"))
```

```{r}
# Load
rawNaive <- read_excel(file.path(dataDir, "GSE100035_Supplementary_Table_1.xlsx"))
rawInjured <- read_excel(file.path(dataDir, "GSE100035_Supplementary_Table_2.xlsx"))
```

## Preview raw gene counts {.tabset}

### Raw
```{r echo=FALSE}
kable(rawNaive[1:2, ], format="html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
```

### Injured
```{r echo=FALSE}
kable(rawInjured[1:2, ], format="html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
```

## {-}

These dataframes indicate the number of times each gene is counted in each sample. `rawNaive` has gene counts for the naive mice. `rawInjured` has gene counts for the injured mice.

- `gene_short_name` contains gene names that might be familiar to a biologist. However, some genes have multiple names and are generally not the most robust. 
- `gene_id` a more robust way of uniquely naming genes. The nomenclature used for these genes comes from Ensemble/Gencode.
- `tss_id` indicates  which transcripts are contributing to a gene's counts; for this analysis, it is not important. 
- `locus` the specific genomic coordinates of a gene. This a handy column to have, but not relevant for this tutorial.
- `avg.FPKM`, `expressed?`, `sum_male`, and `sum_female` are calculated by the data's original authors, and is not relevant for this tutorial.
- `S209_male` to `S230_female` are columns representing gene counts in each sample, where each column is a different sample. Similarlty, `SNL_male` to `SNL_female_260144` in the second dataset each represent a sample.

## Download and load sample metadata
```{r include=FALSE}
# Download 
system(paste('wget -P', dataDir, "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100035/matrix/GSE100035_series_matrix.txt.gz"))
# Unzip
system(paste('gunzip', file.path(dataDir, "GSE100035_series_matrix.txt.gz")))
```

```{r error=FALSE, message=FALSE, warning=FALSE}
# Load metadata about samples into an `ExpressionSet` object
seriesEset <- getGEO(filename = file.path(dataDir, "GSE100035_series_matrix.txt"))
# Extract sample metadata
seriesMatrix <- seriesEset@phenoData@data
```

### Preview

Excluding columns with large amounts of text for the sake of visuals

```{r echo=FALSE}
seriesMatrix %>% 
  dplyr::select(-c(17:18, 22:31, "description.1", "status", "characteristics_ch1.1")) %>% 
  head(n=2) %>% 
  kable(format="html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
```

The columns of interest for us are, `characteristics_ch1.2`, `characteristics_ch1.3`, and `description`.

- `characteristics_ch1.2` indicates the sex of samples
- `characteristics_ch1.3` indicates the injury status of samples
- `description` contains sample IDs that match the column names of the gene counts excel sheet

# Clean up raw variables for analysis

Let's select only the variables that we care about from the count dataframes and seriesMatrix variable.
Note that we never overwrite a variable! Overwriting variables is a great way to get unreliable results.

We use a pipe operator %>% to take the output of one function and feed it as the first argument to a next function, without creating a new variable. I find pipes generally make your code easier to read and help your global environment stay clean.

## Metadata 

Extracting the columns of interest that were previously mentioned.

Rather than storing the metadata in a variable called "metadata", I prefer to call it "coldata". This choice is purely stylistic and personal preference. I mention this caveat because I found it confusing to see others using a variable called "coldata" when I first started learning DEA.
```{r}
coldata <- seriesMatrix %>% 
  # Select variables
  dplyr::select(c(description, characteristics_ch1.2, characteristics_ch1.3)) %>% 
  # Make column names more descriptive
  dplyr::rename(
    sampleId = description, 
    sex = characteristics_ch1.2, 
    phenotype = characteristics_ch1.3) %>% 
  # Cleanup values in columns
  mutate(
    # Remove "Sex: " from `characteristics_ch1.2` column using regex
    sex = gsub("Sex: ", "", sex), 
    # Remove "injury condition: " from `characteristics_ch1.2` column using regex
    phenotype = gsub("injury condition: ", "", phenotype)
    ) %>% 
  # Remove special characters from the injury column 
  #   If value in phenotype is "naïve", remove the special character. Else, replace the value with "injured"
  mutate(phenotype = ifelse(
    phenotype == "naïve", "naive", "injured"
  )) %>% 
  # Ensure the `phenotype` and `sex` columns are stored as a factor
  mutate(
    sex = factor(sex, levels = c("male", "female")), 
    phenotype = factor(phenotype, levels = c("naive", "injured"))
    )
rownames(coldata) <- coldata$sampleId
```

For reasons that will be explained, split coldata by phenotype
```{r}
coldataList <- list()
for (pheno in unique(coldata$phenotype)){
  coldataList[[pheno]] <- coldata %>% subset(phenotype == pheno)
}
```

### Preview {.tabset}

#### Naive
```{r echo=FALSE}
kable(coldataList$naive[1:2, ], format="html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
```

#### Injured
```{r echo=FALSE}
kable(coldataList$injured[1:2, ], format="html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
```

### {-}

## Gene counts

While it may be tempting to merge the naive and injured datasets, it appears that this may be inappropriate. First, the rawInjured and rawNaive datasets have a different number of rows/genes, suggesting that they were processed separately. This is important, because  processing samples separately can cause unwanted technical variation and skew the results. Hence, we will not be able to directly compare naive mice to injured mice.

Let's replace the row numbers with gene ids.
```{r error=TRUE}
rawNaive %>% 
  tibble::column_to_rownames("gene_id")
```
This fails because at least one gene id occurs multiple times! I suspect this is because different transcripts are being attributed to the same gene, and those transcripts are not being totally merged. Ex: 
```{r}
rawNaive %>% 
  group_by(gene_id) %>% 
  filter(n() > 1)
```
Given that the original authors do not clearly address this issue in their manuscript, we will use the `make.names()` function to append ".1" to end of gene_id duplicants. Other options would be to 1) pick the gene with the highest average counts, or 2) calculate the average of both genes' counts. Ideally, I would make sure the previous data processing steps were completed correctly, but that is beyond the scope of this tutorial.

Since the code to execute this data cleaning is not trivially short and needs to run twice, let's create a function!
```{r}
GenesToRownames <- function(inDf){
  #' This function to  cleanup gene_ids for a gene count matrix
  #' @param inDf An input data frame with gene counts
  #' @return returns a data frame with gene counts. Row names are gene ids
  outDf <- inDf %>% 
    # Fix duplicants
    mutate("unique_gene_id" = make.names(gene_id, unique = TRUE)) %>% 
    # Remove the gene_id column
    dplyr::select(-c(gene_id)) %>% 
    # Replace rownames with gene ids
    tibble::column_to_rownames("unique_gene_id")
  # Return
  return(outDf)
}
```

Next, let's remove the columns that we don't need, such as `gene_short_name`, `tss_id`, `locus`, etc. We will also need to make sure that the order of samples in the gene counts matches the order of samples in the coldatas. Let's achieve both aims at once with a function!
```{r}
OrganizeGeneColumns <- function(inCounts, inColdat){
  #' This function selects only sample columns and reorders them to match coldata's rows.
  #' @param inCounts dataframe with gene counts
  #' @param inColdat dataframe with sample metadata
  #' @return Returns an organized count dataframe
  outCounts <- inCounts %>% 
    dplyr::select(as.character(inColdat$sampleId))
  return(outCounts)
}
```

Now let's execute the functions!
```{r}
naiveCounts <- rawNaive %>% GenesToRownames() %>% OrganizeGeneColumns(inColdat = coldataList$naive)
injuredCounts <- rawInjured %>% GenesToRownames() %>% OrganizeGeneColumns(inColdat = coldataList$injured)
# Preview
head(naiveCounts, n = 2)
head(injuredCounts, n = 2)
```

# Quality Control

Let's do some very basic quality control of samples by using unsupervised clustering. Specifically, we will do Principal Component Analysis with DESeq2.

> Input for DESeq2 should always be raw counts. Based on author comments, it appears that the counts downloaded are FPKM transformed, i.e. not raw. For this reason, the code used in this tutorial is fine, but the results should be taken with a grain of salt.

## Initalize DESeq2 Objects

Our variable of interest for DEA is sex. I.e. what genes are different between male and female?
DESeq2 expects gene counts to be whole numbers. Since our counts are floats, we simply round them.

```{r}
ddsNaive <- DESeqDataSetFromMatrix(
  countData = round(naiveCounts), colData = coldataList$naive, design = ~sex)
ddsInjured <- DESeqDataSetFromMatrix(
  countData = round(injuredCounts), colData = coldataList$injured, design = ~sex)
```

## Transformation

The purpose of transforming gene counts prior to unsupervised clustering is to prevent highly expressed genes from nominating the clustering results. DESeq2 provides two methods to achieve this: 

1) Regularized log transformation (`rlog`): Proportedly more robust size factors, calculated from sample gene composition and sequencing depth, varies widely.
2) Variance stabilizing transformation (`vst`): Less computationally expensive, so used when the input dataset is very large.

Let's try both and compare! 
```{r}
DoTransformations <- function(ddsObj){
  #' This function performs both the rlog and vst transformation and returns as a named list
  tList <- list(
    "rlog" = rlog(ddsObj), 
    "vst" = varianceStabilizingTransformation(ddsObj)
  )
  return(tList)
}
t.ddsNaive <- DoTransformations(ddsNaive)
t.ddsInjured <- DoTransformations(ddsInjured)
```

## PCA Calculations

Storing results in a named list, so we don't have to copy-and-paste as much code during the plotting step.
```{r}
pcaDataList <- list(
  "rlog.naive" = plotPCA(t.ddsNaive$rlog, intgroup = "sex", returnData = TRUE),
  "rlog.injured" = plotPCA(t.ddsInjured$rlog, intgroup = "sex", returnData = TRUE),
  "vst.naive" = plotPCA(t.ddsNaive$vst, intgroup = "sex", returnData = TRUE),
  "vst.injured" = plotPCA(t.ddsInjured$vst, intgroup = "sex", returnData = TRUE)
)
```

## PCA Plots
```{r}
# Generate plots and store in list
pcaPlots <- list()
for (set in names(pcaDataList)){
  pcaPlots[[set]] <- pcaDataList[[set]] %>% 
    ggplot(aes(x = PC1, y = PC2, colour = sex, shape = sex)) + 
  geom_point(size = 3) + 
  ggtitle(set) + 
  theme_bw() + 
  coord_fixed()
}
# Show all plots with cowplot
plot_grid(pcaPlots[[1]] + theme(legend.position = "none"), pcaPlots[[2]], 
          pcaPlots[[3]] + theme(legend.position = "none"), pcaPlots[[4]], 
          rel_widths = c(2, 1.75))
```

Albiet the clustering is different between rlog and vst, but the gist is the same. The PCAs don't reveal any clear distinction between the male and female samples; perhaps we won't see many differentially expressed genes male vs. female. Also, you might argue that the two naive male samples that cluster further away from the other samples should be removed, but I think they're not distinct enough to be removed.

# DEA with DESeq2

Let's begin differential expression analysis with DESeq2! 

## Estimating size factors for normalization
```{r}
sf.ddsNaive <- estimateSizeFactors(ddsNaive)
sf.ddsInjured <- estimateSizeFactors(ddsInjured)
```

## Non-specific filtering

Use plots and distributions of MAD to figure out what a good filtering threshold is. Filtering after normalization and dispersion estimation to prevent estimations and shrinkage from getting calculated incorrectly.
MAD = median absolute deviation, more robust than something like standard deviation

### Functions
```{r}
GetMads <- function(counts){
  # Calculate the median absolute deviation of each gene/transcript across samples
  # Input: counts dataframe
  # Output: dataframe with each gene's MAD
  mads <- data.frame(apply(X = counts, MARGIN = 1,  FUN = mad))
  colnames(mads) <- "mad"
  return(mads)
}
GetMadCutoff <- function(counts){
  # Visualize quantiles to determine the optimal cutoff
  # Input: feature (gene/transcript) counts
  # Output: saves a plot called "madQuantiles.pdf" in the plotsDir; returns the MADs
  madsAll <- data.frame(id = rownames(counts), mad = GetMads(counts)) 
  madsQuantiles <- quantile(madsAll$mad, probs = seq(0, .9, by = .01))
  # Figure out how many features are kept after each filtering cutoff
  madNumKept <- c()
  quantileLables <- c()
  for (val in madsQuantiles){
    madNumKept <- append(madNumKept, length(subset(madsAll, mad >= val)$mad))
    quantileLables <- append(quantileLables, names(madsQuantiles)[match(val, madsQuantiles)])
  }
  plotMadQuantiles <- data.frame(quantileLables, madNumKept)
  # Make plot
  gplot <- plotMadQuantiles %>% 
    ggplot(aes(x = as.numeric(gsub("%", "", quantileLables)), y = madNumKept)) + 
    geom_point() + 
    scale_x_continuous(name = "Percentile", breaks = seq(0, 100, 10), limits = c(0, 100)) + 
    ylab("Number of remaining features") + # features = gene or transcript
    ggtitle("Visualizing MAD cutoffs") + 
    theme_bw() +
    theme(plot.title = element_text(hjust = 0.5))
  return(list("madsAll" = madsAll, "gplot" = gplot))
} 
```

### Execute
```{r}
mads <- GetMadCutoff(counts(sf.ddsNaive, normalized = TRUE))
mads[["gplot"]]
```

Based on the plot, choose range of quantiles to search for an exact threshold
```{r}
quantile(mads$mad, probs = seq(.5, 0.7, 0.1)) # adjust range in seq until see a transition between 0 and 1
```

Remove lowly variable genes
```{r}
filt_threshold <- 0.006548972 # based on this distribution and plot, this MAD is a good cutoff 
toKeep <- rownames(subset(mads, mad >= filt_threshold))
filt.ddsNaive <- sf.ddsNaive[toKeep, ]
filt.ddsInjured <- sf.ddsInjured
```

### What are the differences?
```{r}
cat(cat("Before filtering: ", nrow(sf.ddsNaive), "\n"), 
    cat("After filtering: ", nrow(filt.ddsNaive), "\n"))
```

## Estimate dispersion to minimize lowly expressed, variable genes
```{r}
ed.ddsNaive <- estimateDispersions(filt.ddsNaive)
ed.ddsInjured <- estimateDispersions(filt.ddsInjured)
```

Let's inspect 
```{r}
plotDispEsts(ed.ddsNaive,
  ylab = "Dispersion", xlab = "Mean of normalized counts",
  main = "Naive"
)
plotDispEsts(ed.ddsInjured,
  ylab = "Dispersion", xlab = "Mean of normalized counts",
  main = "Injured"
)
```

## Wald test
```{r}
wt.ddsNaive <- nbinomWaldTest(ed.ddsNaive)
wt.ddsInjured <- nbinomWaldTest(ed.ddsInjured)
```

## Get results
```{r}
res.Naive <- results(wt.ddsNaive, contrast=c("sex","female","male"), 
                     alpha = 0.05, pAdjustMethod = "BH")
res.Injured <- results(wt.ddsInjured, contrast=c("sex","female","male"), 
                     alpha = 0.05, pAdjustMethod = "BH")
```

### Convert results to dataframes
```{r}
# Naive
allRes.Naive <- res.Naive %>% as.data.frame() %>% 
  arrange(padj) %>% 
  tibble::rownames_to_column("unique_gene_id") %>% 
  # remove our artificial .1's from gene ids
  mutate(gene_id = gsub("\\.1", "", unique_gene_id)) %>% 
  dplyr::select(-c(unique_gene_id)) %>% 
  # add gene information from raw counts
    # within left_join, selecting columsn from rawNaive, 
    # and ensuring all rows are unique
  left_join(rawNaive %>% 
              select(c(gene_id, gene_short_name)) %>% 
              unique(), 
            by = join_by(gene_id)) %>% 
  # move gene columns to the front
  relocate(gene_id, gene_short_name)
# Injured
allRes.Injured <- res.Injured %>% as.data.frame() %>% 
  arrange(padj) %>% 
  tibble::rownames_to_column("unique_gene_id") %>% 
  mutate(gene_id = gsub("\\.1", "", unique_gene_id)) %>% 
  dplyr::select(-c(unique_gene_id)) %>% 
  left_join(rawInjured %>% 
              select(c(gene_id, gene_short_name)) %>% 
              unique(), 
            by = join_by(gene_id)) %>% 
  relocate(gene_id, gene_short_name)
```

## Subset by significant
```{r}
sigRes.Naive <- allRes.Naive %>% subset(padj < 0.05)
sigRes.Injured <- allRes.Injured %>% subset(padj < 0.05)
```

## Summarize results
```{r}
cat("For the naive dataset,", nrow(sigRes.Naive), "genes were differentially expressed (BH adjusted p-value < 0.05), out of the", nrow(allRes.Naive), "tested genes.")

cat("\n\nFor the injured dataset,", nrow(sigRes.Injured), "genes were differentially expressed (BH adjusted p-value < 0.05), out of the", nrow(allRes.Injured), "tested genes.")
```
## Preview {.tabset}

### Naive
```{r echo=FALSE}
kable(sigRes.Naive, format="html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
```

### Injured
```{r echo=FALSE}
kable(sigRes.Injured, format="html") %>% kable_styling("striped") %>% scroll_box(width = "100%")
```

## {-}

# Plot results









